src/ir/predicate_test.cc

```diff
@@ -86,73 +93,118 @@ INSTANTIATE_TEST_SUITE_P(ToDatalogRuleBodyTest, ToDatalogRuleBodyTest,
-static const And kAndTag2Tag3(std::make_unique<TagPresence>(kTag2Present),
-                             std::make_unique<TagPresence>(kTag3Present));
-
-static const Or kOrTag1Tag2(std::make_unique<TagPresence>(kTag1Present),
-                             std::make_unique<TagPresence>(kTag2Present));
-static const Or kOrTag1Tag3(std::make_unique<TagPresence>(kTag1Present),
-                             std::make_unique<TagPresence>(kTag3Present));
-static const Or kOrTag2Tag3(std::make_unique<TagPresence>(kTag2Present),
-                             std::make_unique<TagPresence>(kTag3Present));
-
-static const Implies kImpliesTag1Tag2(
-    std::make_unique<TagPresence>(kTag1Present),
-    std::make_unique<TagPresence>(kTag2Present));
-static const Implies kImpliesTag1Tag3(
-    std::make_unique<TagPresence>(kTag1Present),
-    std::make_unique<TagPresence>(kTag3Present));
-static const Implies kImpliesTag2Tag3(
-    std::make_unique<TagPresence>(kTag2Present),
-    std::make_unique<TagPresence>(kTag3Present));
-
-static const Not kNotTag1(std::make_unique<TagPresence>(kTag1Present));
-static const Not kNotTag2(std::make_unique<TagPresence>(kTag2Present));
-static const Not kNotTag3(std::make_unique<TagPresence>(kTag3Present));
-
-static const Predicate *example_predicate_pointers[] = {
-    &kTag1Present,
-    &kTag2Present,
-    &kTag3Present,
-    &kAndTag1Tag2,
-    &kAndTag1Tag3,
-    &kAndTag2Tag3,
-    &kOrTag1Tag2,
-    &kOrTag1Tag3,
-    &kOrTag2Tag3,
-    &kImpliesTag1Tag2,
-    &kImpliesTag1Tag3,
-    &kImpliesTag2Tag3,
-    &kNotTag1,
-    &kNotTag2,
-    &kNotTag3,
+static const Predicate *kTag1Present = TagPresence::Create(kArena, "tag1");
+static const Predicate *kTag2Present = TagPresence::Create(kArena, "tag2");
+static const Predicate *kTag3Present = TagPresence::Create(kArena, "tag3");
+
+static const Predicate *kTag1Present2 = TagPresence::Create(kArena, "tag1");
+static const Predicate *kTag2Present2 = TagPresence::Create(kArena, "tag2");
+static const Predicate *kTag3Present2 = TagPresence::Create(kArena, "tag3");
+
+static const And *kAndTag1Tag2 =
+    And::Create(kArena, kTag1Present, kTag2Present);
+static const And *kAndTag1Tag3 =
+    And::Create(kArena, kTag1Present, kTag3Present);
+static const And *kAndTag2Tag3 =
+    And::Create(kArena, kTag2Present, kTag3Present);
+
+static const And *kAndTag1Tag2_2 =
+    And::Create(kArena, kTag1Present, kTag2Present);
+static const And *kAndTag1Tag3_2 =
+    And::Create(kArena, kTag1Present, kTag3Present);
+static const And *kAndTag2Tag3_2 =
+    And::Create(kArena, kTag2Present, kTag3Present);
+
+static const Or *kOrTag1Tag2 = Or::Create(kArena, kTag1Present, kTag2Present);
+static const Or *kOrTag1Tag3 = Or::Create(kArena, kTag1Present, kTag3Present);
+static const Or *kOrTag2Tag3 = Or::Create(kArena, kTag2Present, kTag3Present);
+
+static const Or *kOrTag1Tag2_2 = Or::Create(kArena, kTag1Present, kTag2Present);
```

Comment:
Why do we need this in addition to `kOrTag1Tag2`? (Same question for other variables as well.)

Reply:
This checks that we have the right behavior for values that are structurally equal, but not pointer equal. My earlier draft of this test created a bunch of structurally-distinct objects and ensured that they only compared equal if they were pointer-equal. I realized that, because I early-exist on pointer equality when comparing predicates, I wasn't actually exercising the case where structural equality compares to true. These duplicates and the numbered equivalence classes fix that issue with the test.